k8s/src/main/scala/io/buoyant/k8s/IngressCache.scala

```diff
@@ -0,0 +1,144 @@
+package io.buoyant.k8s
+
+import com.twitter.finagle.Service
+import com.twitter.finagle.http.{Request, Response}
+import com.twitter.util._
+import java.util.concurrent.atomic.AtomicReference
+
+case class IngressSpec(
+  name: Option[String],
+  namespace: Option[String],
+  fallbackBackend: Option[IngressPath] = None,
+  rules: Seq[IngressPath]
+) {
+  def getMatchingPath(hostHeader: Option[String], requestPath: String): Option[IngressPath] = {
+    val matchingPath = rules.find(_.matches(hostHeader, requestPath))
+    (matchingPath, fallbackBackend) match {
+      case (Some(path), _) =>
+        log.info("k8s found rule matching %s %s: %s", hostHeader.getOrElse(""), requestPath, path)
+        Some(path)
+      case (None, Some(default)) =>
+        log.info("k8s using default service %s for request %s %s", default, hostHeader.getOrElse(""), requestPath)
+        Some(default)
+      case _ =>
+        log.info("k8s no suitable rule found in %s for request %s %s", name.getOrElse(""), hostHeader.getOrElse(""), requestPath)
+        None
+    }
+  }
+}
+
+case class IngressPath(
+  host: Option[String] = None,
+  uri: Option[String] = None,
+  namespace: String,
+  svc: String,
+  port: String
+) {
+  def uriMatches(uri: String, path: String): Boolean = path.isEmpty || uri.matches(path)
```

Comment:
for performance reasons we may want to compile this regex and store it as a val on the class.